lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -583,6 +818,180 @@ pub unsafe extern "C" fn handle_tdp_sd_info_response(
     }
 }
 
+/// handle_tdp_sd_create_response handles a TDP Shared Directory Create Response
+/// message
+///
+/// # Safety
+///
+/// client_ptr MUST be a valid pointer.
+/// (validity defined by https://doc.rust-lang.org/nightly/core/primitive.pointer.html#method.as_ref-1)
+#[no_mangle]
+pub unsafe extern "C" fn handle_tdp_sd_create_response(
+    client_ptr: *mut Client,
+    res: CGOSharedDirectoryCreateResponse,
+) -> CGOErrCode {
+    // # Safety
+    //
+    // This function MUST NOT hang on to any of the pointers passed in to it after it returns.
+    // All passed data that needs to persist after this function MUST be copied into Rust-owned memory.
+
+    // Clone here to ensure nothing the CGO object is passed to can hang on to it or any of its pointers.
+    #[allow(clippy::redundant_clone)]
+    let res: SharedDirectoryCreateResponse = res.clone();
+
+    let client = match Client::from_ptr(client_ptr) {
+        Ok(client) => client,
+        Err(cgo_error) => {
+            return cgo_error;
+        }
+    };
+
+    let mut rdp_client = client.rdp_client.lock().unwrap();
+    match rdp_client.handle_tdp_sd_create_response(res) {
+        Ok(()) => CGOErrCode::ErrCodeSuccess,
+        Err(e) => {
+            error!("failed to handle Shared Directory Create Response: {:?}", e);
+            CGOErrCode::ErrCodeFailure
+        }
+    }
+}
+
+/// handle_tdp_sd_delete_response handles a TDP Shared Directory Delete Response
+/// message
+///
+/// # Safety
+///
+/// client_ptr MUST be a valid pointer.
+/// (validity defined by https://doc.rust-lang.org/nightly/core/primitive.pointer.html#method.as_ref-1)
+#[no_mangle]
+pub unsafe extern "C" fn handle_tdp_sd_delete_response(
+    client_ptr: *mut Client,
+    res: CGOSharedDirectoryDeleteResponse,
+) -> CGOErrCode {
+    // # Safety
+    //
+    // This function MUST NOT hang on to any of the pointers passed in to it after it returns.
+    // All passed data that needs to persist after this function MUST be copied into Rust-owned memory.
+
+    // Clone here to ensure nothing the CGO object is passed to can hang on to it or any of its pointers.
```

Comment:
I don't quite understand these comments and why the clone is necessary. What's going on here?